src/main/java/com/google/api/gax/grpc/AbstractPage.java

```diff
@@ -29,76 +29,81 @@
+
+  protected interface PageFactory<RequestT, ResponseT, ResourceT, PageT> {
+    PageT createPage(
+        UnaryCallable<RequestT, ResponseT> callable,
+        PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+        RequestT request,
+        CallContext context,
+        ResponseT response);
+  }
 
   private final UnaryCallable<RequestT, ResponseT> callable;
   private final PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor;
   private final RequestT request;
   private final CallContext context;
   private final ResponseT response;
 
-  public PageContext(
+  protected static <RequestT, ResponseT, ResourceT, PageT> PageT callApiAndCreate(
+      PageFactory<RequestT, ResponseT, ResourceT, PageT> factory,
       UnaryCallable<RequestT, ResponseT> callable,
       PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
       RequestT request,
       CallContext context) {
+    ResponseT response =
+        ApiExceptions.callAndTranslateApiException(callable.futureCall(request, context));
+    return factory.createPage(callable, pageDescriptor, request, context, response);
+  }
+
+  protected AbstractPage(
+      UnaryCallable<RequestT, ResponseT> callable,
+      PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+      RequestT request,
+      CallContext context,
+      ResponseT response) {
     this.callable = callable;
     this.pageDescriptor = pageDescriptor;
     this.request = request;
     this.context = context;
-
-    // Make the API call eagerly
-    this.response =
-        ApiExceptions.callAndTranslateApiException(callable.futureCall(request, context));
-  }
-
-  public Iterable<ResourceT> getResourceIterable() {
-    return pageDescriptor.extractResources(response);
+    this.response = response;
   }
 
+  @Override
   public boolean hasNextPage() {
     return !getNextPageToken().equals(pageDescriptor.emptyToken());
   }
 
+  @Override
   public String getNextPageToken() {
     return pageDescriptor.extractNextToken(response);
   }
 
-  public PageContext<RequestT, ResponseT, ResourceT> getNextPageContext() {
-    if (!hasNextPage()) {
-      return null;
-    }
-    RequestT nextRequest = pageDescriptor.injectToken(request, getNextPageToken());
-    return new PageContext<>(callable, pageDescriptor, nextRequest, context);
-  }
-
-  public PageContext<RequestT, ResponseT, ResourceT> getNextPageContext(int pageSize) {
-    if (!hasNextPage()) {
-      return null;
-    }
-    RequestT nextRequest = pageDescriptor.injectToken(request, getNextPageToken());
-    nextRequest = pageDescriptor.injectPageSize(nextRequest, pageSize);
-    return new PageContext<>(callable, pageDescriptor, nextRequest, context);
-  }
-
-  public int getPageElementCount() {
-    return Iterables.size(getResourceIterable());
-  }
+  @Override
+  public abstract AbstractPage<RequestT, ResponseT, ResourceT> getNextPage();
```

Comment:
It can't be concrete, because AbstractPage can't be instantiated. It could just be omitted.